File: Asad-sAIPF-main/scratch/upscale.js
const fs = require('fs');
const path = require('path');

async function upscale() {
  try {
    let sharp;
    try {
      sharp = require('sharp');
    } catch (e) {
      console.log('Installing sharp for HD image processing...');
      require('child_process').execSync('npm install sharp --no-save', { stdio: 'inherit' });
      sharp = require('sharp');
    }
    
    const imgPath = path.join(__dirname, '../public/portrait.png');
    const tempPath = path.join(__dirname, '../public/portrait_temp.png');
    
    if (!fs.existsSync(imgPath)) {
      console.log('Image not found at:', imgPath);
      return;
    }
    
    console.log('Reading portrait.png...');
    const metadata = await sharp(imgPath).metadata();
    console.log(`Original resolution: ${metadata.width}x${metadata.height}`);
    
    // Upscale by 3x using Lanczos3 high-definition resampling kernel
    const newWidth = Math.round(metadata.width * 3);
    const newHeight = Math.round(metadata.height * 3);
    console.log(`Upscaling to: ${newWidth}x${newHeight} using Lanczos3 and Unsharp Masking...`);
    
    await sharp(imgPath)
      .resize({
        width: newWidth,
        height: newHeight,
        kernel: sharp.kernel.lanczos3,
        withoutEnlargement: false
      })
      .sharpen({
        sigma: 2.0,
        m1: 1.8,
        m2: 0.8,
        x1: 2,
        y2: 10,
        y3: 20
      })
      .modulate({
        brightness: 1.02,
        saturation: 1.08
      })
      .png({
        quality: 100,
        compressionLevel: 9
      })
      .toFile(tempPath);
      
    // Overwrite original with HD version
    fs.renameSync(tempPath, imgPath);
    console.log('✨ SUCCESS! Your portrait image has been 3x upscaled and sharpened to ultra-HD studio quality!');
  } catch (err) {
    console.error('Error upscaling:', err.message);
  }
}

upscale();

File: Asad-sAIPF-main/scratch/upscale.py
import os
try:
    from PIL import Image, ImageEnhance
except ImportError:
    import subprocess
    import sys
    subprocess.check_call([sys.executable, "-m", "pip", "install", "pillow"])
    from PIL import Image, ImageEnhance

img_path = r"x:\Self projects\AsadPFW272026\public\portrait.png"
if os.path.exists(img_path):
    print(f"Opening image: {img_path}")
    img = Image.open(img_path)
    print(f"Original resolution: {img.size[0]}x{img.size[1]}")
    
    # Convert to RGBA to preserve transparency
    img = img.convert("RGBA")
    
    # Upscale by 2.5x using Lanczos high-definition resampling
    new_width = int(img.width * 2.5)
    new_height = int(img.height * 2.5)
    upscaled = img.resize((new_width, new_height), Image.Resampling.LANCZOS)
    
    # Apply Unsharp Masking / Sharpness Enhancement
    sharpener = ImageEnhance.Sharpness(upscaled)
    sharpened = sharpener.enhance(2.3)  # Boost sharpness cleanly
    
    # Boost contrast slightly
    contraster = ImageEnhance.Contrast(sharpened)
    enhanced = contraster.enhance(1.08)
    
    # Boost color vibrancy slightly
    colorer = ImageEnhance.Color(enhanced)
    final_img = colorer.enhance(1.05)
    
    # Save back to disk with maximum PNG quality
    final_img.save(img_path, "PNG", quality=100, optimize=True)
    print(f"Successfully enhanced and upscaled image to {final_img.size[0]}x{final_img.size[1]}!")
else:
    print(f"Image not found at {img_path}")
